### PR TITLE
Brunch integration with phoenix.new options

### PR DESCRIPTION
**Pull request by chrismccord**

### CODE REVIEW

**Comment by chrismccord** [comment]
So the file handling for the brunch/no brunch cases is kind of a mess and I could use some feedback on making it cleaner. There's overlap on what files need to go where `priv/static` or `web/static` and either case requires copying or removing directories from the template. The new options are the following:

    # Do not include any brunch related files (package.json, brunch-config.js, web/static/)
    mix phoenix.new /path/to --skip-brunch
    # include brunch, but not automatically run `npm install` in project
    mix phoenix.new /path/to --skip-npm

**Comment by chrismccord** [comment]
I also just realized, we need to conditionally generate the template/config/dev.exs to exclude `watches: []`

File: lib/mix/tasks/phoenix.new.ex [hunk 1]
```diff
@@ -42,30 +51,73 @@ defmodule Mix.Tasks.Phoenix.New do
         mod = opts[:module] || Naming.camelize(app)
         check_mod_name!(mod)
 
-        run(app, mod, path, opts[:dev])
+        run(app, mod, path, opts[:dev], opts[:skip_brunch], opts[:skip_npm])
```
Comment (josevalim) [inline lib/mix/tasks/phoenix.new.ex]: At this point we should just pass all `opts`? Don't worry though, I can refactor it when working on the model generator.

File: lib/phoenix/endpoint/adapter.ex [hunk 2]
```diff
@@ -4,26 +4,45 @@ defmodule Phoenix.Endpoint.Adapter do
...
+    pub_conf   = conf[:pubsub]
 
     pubsub_children = case pub_conf[:adapter] do
       nil     -> []
       adapter -> [supervisor(adapter, [mod.__pubsub_server__(), pub_conf])]
     end
 
-    children = pubsub_children ++ [
+    watcher_children =
+      Enum.reduce(conf[:watchers], [], fn {cmd, args}, acc ->
```
Comment (josevalim) [inline lib/phoenix/endpoint/adapter.ex]: Isn't this a map?

Comment (josevalim) [inline lib/phoenix/endpoint/adapter.ex]: I mean, `Enum.map/2`.

File: lib/phoenix/endpoint/adapter.ex [hunk 3]
```diff
@@ -4,26 +4,45 @@ defmodule Phoenix.Endpoint.Adapter do
...
+      paths -> [worker(ChangeDetector, [paths, {__MODULE__, :assets_change, [mod]}])]
+    end
+
+    children = watcher_children ++ live_reload_children ++ pubsub_children ++ [
       worker(Phoenix.Config, [otp_app, mod, defaults(otp_app, mod)]),
-      supervisor(Phoenix.Endpoint.Server, [otp_app, mod])
+      supervisor(Phoenix.Endpoint.Server, [otp_app, mod]),
     ]
 
     Supervisor.start_link(children, strategy: :rest_for_one, name: mod)
```
Comment (josevalim) [inline lib/phoenix/endpoint/adapter.ex]: It needs to definitely be `:one_for_one`.

File: priv/static/phoenix.js.map [hunk 4]
```diff
@@ -0,0 +1 @@
...\n        case 0:\n        case 500:\n          this.onerror()\n          this.closeAndRetry()\n          break\n        default: throw(`unhandled poll status ${status}`)\n      }\n    })\n  }\n\n  send(body){\n    Ajax.request(\"POST\", this.endpointURL(), \"application/json\", body, this.onerror.bind(this, \"timeout\"), (status, resp) => {\n      if(status !== 200){ this.onerror(status) }\n    })\n  }\n\n  close(code, reason){\n    this.readyState = this.states.closed\n    this.onclose()\n  }\n}\n\n\nexport var Ajax = {\n\n  states: {complete: 4},\n\n  request: function(method, endPoint, accept, body, ontimeout, callback){\n    let req = XMLHttpRequest ? new XMLHttpRequest() : // IE7+, Firefox, Chrome, Opera, Safari\n                               new ActiveXObject(\"Microsoft.XMLHTTP\") // IE6, IE5\n    req.open(method, endPoint, true)\n    req.setRequestHeader(\"Content-type\", accept)\n    req.onerror = function(){ callback && callback(500, null) }\n    req.onreadystatechange = () => {\n      if(req.readyState === this.states.complete && callback){ callback(req.status, req.responseText) }\n    }\n    if(ontimeout){ req.ontimeout = ontimeout }\n\n    req.send(body)\n  }\n}\n"]}
```
Comment (josevalim) [inline priv/static/phoenix.js.map]: Do we need this .map file? Shouldn't the compiler phoenix.js already be minified? I.e. compiled in production mode?

File: priv/template/config/dev.exs [hunk 5]
```diff
@@ -3,7 +3,12 @@ use Mix.Config
 config :<%= application_name %>, <%= application_module %>.Endpoint,
   http: [port: System.get_env("PORT") || 4000],
   debug_errors: true,
-  cache_static_lookup: false
+  cache_static_lookup: false,
+  watchers: [{Path.expand("node_modules/brunch/bin/brunch"), ["watch"]}],
+  live_reload: ["priv/static/app.js",
+                "priv/static/app.css",
+                Path.wildcard("web/templates/**/*")]
```
Comment (josevalim) [inline priv/template/config/dev.exs]: I thought we were not going to ship with template watching by default?

File: priv/template/lib/application_name/endpoint.ex [hunk 6]
```diff
@@ -4,7 +4,8 @@ defmodule <%= application_module %>.Endpoint do
   # Serve at "/" the given assets from "priv/static" directory
   plug Plug.Static,
     at: "/", from: :<%= application_name %>,
-    only: ~w(css images js favicon.ico robots.txt)
+    only: ~w(css images js favicon.ico robots.txt
```
Comment (josevalim) [inline priv/template/lib/application_name/endpoint.ex]: I though we agreed to compile everything into `css` and `js` so we don't need the extra line below?

File: lib/mix/tasks/phoenix.new.ex (hunk 1 above)
Comment (chrismccord) [inline lib/mix/tasks/phoenix.new.ex]: :+1: agree

File: lib/phoenix/endpoint/adapter.ex (hunk 2 above)
Comment (chrismccord) [inline lib/phoenix/endpoint/adapter.ex]: Yes. Initially I was joining the live reload + watchers as one list, so this can just be a map now

**Comment by josevalim** [comment]
I think the generator could be cleaner if you remove all static files out from `priv/template` and keep them in `priv/static`. Then create two maps of how you want them to be copied:

```elixir
@brunch %{
  "phoenix.js" => "vendor/phoenix.js",
  "phoenix.png" => "assets/images/phoenix.png"
  ...
}

@static %{
  "phoenix.js" => "..."
}
```

Now you run copy_from or something similar passing the map:

```elixir
# With brunch
copy_from "priv/static", "web/static", @brunch

# Without
copy_from "priv/static", "priv/static", @static
```

Otherwise today we are copying files and then removing them and it ends up being very confusing. Also, because there is no map in the current implementation and the code is scattared around a couple places, it is very hard for me to look at the code and understand how the target directory is going to look like, be it for brunch, or for static.

File: priv/template/lib/application_name/endpoint.ex (hunk 6 above)
Comment (chrismccord) [inline priv/template/lib/application_name/endpoint.ex]: Ah yes. Updated!

**Comment by chrismccord** [comment]
@josevalim Thanks. That sounds like a really nice solution. I'll make the changes this evening

File: priv/static/phoenix.js.map (hunk 4 above)
Comment (chrismccord) [inline priv/static/phoenix.js.map]: We don't need the map, and yes should be min'd

File: priv/template/config/dev.exs (hunk 5 above)
Comment (chrismccord) [inline priv/template/config/dev.exs]: I think we should include template watching by default, but add a comment along the lines of:

```
  # Watch assets and templates for live-reload changes
  # Note: larger projects will use higher CPU in dev as the number of files grows
```

File: priv/static/phoenix.js.map (hunk 4 above)
Comment (josevalim) [inline priv/static/phoenix.js.map]: :shipit:

**Comment by chrismccord** [comment]
@josevalim I made the discussed changes and it's waaaay nicer. Thanks! Give it a look and I think we're ready!

**Comment by scrogson** [comment]
Looks really good @chrismccord! :+1:

**Pull request merged**